### ISSUE TITLE
Error deploying app from local docker registry

### ISSUE THREAD

**Issue by spencergibb** [cloudfoundry-incubator/lattice#51]
```
$ ltc create myfeed-config 192.168.11.1:5000/myfeed-config
Error fetching image metadata: Error Connecting to Docker registry:
invalid registry endpoint https://192.168.11.1:5000/v0/: unable to ping registry endpoint https://192.168.11.1:5000/v0/
v2 ping attempt failed with error: Get https://192.168.11.1:5000/v2/: EOF
 v1 ping attempt failed with error: Get https://192.168.11.1:5000/v1/_ping: EOF. If this private registry supports only HTTP or HTTPS with an unknown CA certificate, please add `--insecure-registry 192.168.11.1:5000` to the daemon's arguments. In the case of HTTPS, if you have access to the registry's CA certificate, no need for the flag; simply place the CA certificate at /etc/docker/certs.d/192.168.11.1:5000/ca.crt
```

I added `--insecure-registry 192.168.11.1:5000` to docker on `boot2docker` which is what allowed me to push the docker image to the first place.

Does `ltc` run docker commands on the host or the vagrant vm?

**Comment by davidwadden** [comment]
Please note: this simply allows "unsigned" registries to be used.  Repositories that require authentication are still unsupported by `ltc`.

There are two changes that need to be made:

1) whitelist custom registry in boot2docker [source](https://github.com/boot2docker/boot2docker#insecure-registry) (this you figured out)
```bash
$ boot2docker init
$ boot2docker up
$ boot2docker ssh "echo $'EXTRA_ARGS=\"--insecure-registry <YOUR INSECURE HOST>\"' | sudo tee -a /var/lib/boot2docker/profile && sudo /etc/init.d/docker restart"
```

2) whitelist custom registry in garden-linux
```bash
$ cd /lattice/repo/folder
$ vagrant ssh
$ sudo vi /etc/init/garden-linux.conf
(make below edits)
$ sudo stop garden-linux;  sudo start garden-linux
```

Add `-insecureDockerRegistryList` argument to garden-linux (described):

`-insecureDockerRegistryList="": comma-separated list of docker registries to allow connection to even if they are not secure`

and, to your question,  it uses the boot2docker vm to fetch metadata, but the vagrant vm actually runs the docker image itself.  Both need to allow connecting to insecure registries.

**Comment by spencergibb** [comment]
@davidwadden Thanks for this.  Is that documented anywhere?

**Comment by spencergibb** [comment]
I followed your instructions and still get the same error.

My `/etc/init/garden-linux.conf`

```
#!upstart

start on started bootstrap
stop on shutdown
respawn

script
    echo "UPSTART: Trying to start garden-linux - `date --rfc-3339=ns`"

    export $(cat /var/lattice/setup/lattice-environment)

    tee2metron -dropsondeDestination=127.0.0.1:3457 -sourceInstance=$LATTICE_CELL_ID \
    garden-linux \
        -disableQuotas=true \
        -listenNetwork=tcp \
        -listenAddr=0.0.0.0:7777 \
        -externalIP=$GARDEN_EXTERNAL_IP \
        -denyNetworks= \
        -allowNetworks= \
        -bin=/var/lattice/garden/garden-bin \
        -depot=/var/lattice/garden/depot \
        -rootfs=/var/lattice/garden/rootfs \
        -overlays=/var/lattice/garden/overlays \
        -graph=/var/lattice/garden/graph \
        -containerGraceTime=1h \
        -allowHostAccess=true \
        -insecureDockerRegistryList="192.168.11.1:5000" \
        >> /var/lattice/log/garden-linux-service.log 2>&1
end script

post-stop exec sleep 5
```

**Comment by spencergibb** [comment]
Btw, the only reference to `insecureDockerRegistryList` is in `old/main.go`.

https://github.com/cloudfoundry-incubator/garden-linux/search?utf8=%E2%9C%93&q=insecureDockerRegistryList

**Comment by spencergibb** [comment]
ping @davidwadden see continued errors above.
```
$ ltc -v
ltc version v0.2.0-rc.2
```

**Comment by onsi** [comment]
gonna reopen this until we get it resolved.

**Comment by onsi** [comment]
@spencergibb `old` is a red-herring.  Garden's in the midst of a long-running reorg/refactor and code in `old` is still in use (sorry for the confusion).

**Comment by spencergibb** [comment]
@onsi thanks.  Good to know.  Still failing after upgrading to latest version and pull in github repo:
```
$ ltc -v
ltc version v0.2.1
```

**Comment by spencergibb** [comment]
any log or anything from the lattice vagrant box I can provide you?

**Comment by jbayer** [comment]
@spencergibb turns out that the `ltc` binary also needs to be updated when using insecure registries according to brief chat with @onsi , but sending LRP json directly to the receptor api endpoint at receptor.SYSTEMDOMAIN like the default vagrant/virtualbox setting of 192.168.11.11.xip.io should work. if i get that working i'll report back.

**Comment by onsi** [comment]
thanks @jbayer -- we also have a story prioritized high to fix the underlying issue in `ltc`.  hopefully shouldn't be long.

**Comment by jbayer** [comment]
i was able to get this to work by using an insecure registry setting on the docker daemon for boot2docker. here is the summary of my edits for the boot2docker vm:

```
# cat /var/lib/boot2docker/profile
EXTRA_ARGS="--insecure-registry 192.168.59.103:5000"
```

and by using this option for garden-linux in /etc/init/garden-linux.conf on the lattice vm:

```
-insecureDockerRegistryList="192.168.59.103:5000"
```

then i ran the private docker registry on boot2docker:

```
docker run -p 5000:5000 registry
```

then i pulled the lattice-app image from dockerhub and pushed it to my local registry:

```
docker run --rm -i -t -p 8080:8080 -e INSTANCE_INDEX=0 -e PORT=8080 cloudfoundry/lattice-app /lattice-app

# get the image id
docker images
REPOSITORY                        TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
registry                          latest              e33e81d7024c        10 days ago         413.7 MB
cloudfoundry/lattice-app          latest              29d531509fb1        3 months ago        19.23 MB

# tag it to the private registry location
docker tag 29d531509fb1 192.168.59.103:5000/lattice-app

# push it to the private registry
docker push 192.168.59.103:5000/lattice-app
```

now post json referencing the private docker registry. you can see example json for the LRP here:
https://gist.github.com/jbayer/9497299cd00aee5ecefe#file-private-lattice-json-L29
take note of the private docker registry syntax. save that as private-lattice.json somewhere.

open up two ltc shells, i like using veritas for pretty print of the lattice component logs: https://github.com/pivotal-cf-experimental/veritas
```
ltc debug-logs | veritas chug
ltc logs lattice-app
```

now do the posting of the JSON to the receptor and see if you get a 201 and see that the lrp is there:
```
 curl -X POST -d @/Users/jamesbayer/Downloads/private-lattice.json http://receptor.192.168.11.11.xip.io/v1/desired_lrps -vvv
* Hostname was NOT found in DNS cache
*   Trying 192.168.11.11...
* Connected to receptor.192.168.11.11.xip.io (192.168.11.11) port 80 (#0)
> POST /v1/desired_lrps HTTP/1.1
> User-Agent: curl/7.37.1
> Host: receptor.192.168.11.11.xip.io
> Accept: */*
> Content-Length: 866
> Content-Type: application/x-www-form-urlencoded
>
* upload completely sent off: 866 out of 866 bytes
< HTTP/1.1 201 Created
< Content-Length: 0
< Content-Type: text/plain; charset=utf-8
< Date: Tue, 31 Mar 2015 05:56:52 GMT
< X-Cf-Requestid: e22dabaf-e809-4508-733b-e8da82539e96
<
* Connection #0 to host receptor.192.168.11.11.xip.io left intact
pintail:~ jamesbayer$ ltc list
App Name	Instances	DiskMB		MemoryMB	Route
lattice-app	1/1		1024		128		lattice-app.192.168.11.11.xip.io, lattice-app-8080.192.168.11.11.xip.io => 8080
```

if there is a problem, the `ltc debug-logs` shell should show it

**Comment by davidwadden** [comment]
I was able to successfully follow @jbayer 's instructions to deploy apps to a private registry, and it seems `ltc` needs a change to deploy these apps onto Lattice:  (adding the following line seems to work):

```
repositoryInfo.Index.Secure = false
```
to docker_session_factory.go:34

Let me discuss whether this is a reasonable fix and we'll make the change to support this.

Thanks,
David

**Comment by spencergibb** [comment]
@davidwadden maybe a cli switch `--secure-registry=false` (or true if false is the default)

**Comment by davidwadden** [comment]
@spencergibb we just made a change that will retry connecting with the insecure option if `ltc` gets the above error ("... If this private registry supports only HTTP or HTTPS with an unknown CA certificate ...")

**Comment by spencergibb** [comment]
@davidwadden sweet!

**Comment by davidwadden** [comment]
this landed on master with v0.2.2-13-g36d868a.  

closing this issue -- let us know if you have any troubles with it @spencergibb
